### ISSUE TITLE
Wrong MD5 checksum when uploading files

### ISSUE THREAD

**Issue by Alaneor** [balderdashy/skipper-s3#29]
It seems that the adapter (or one of its dependencies) calculates an MD5 checksum of the uploaded file and sends it to S3 along with the file itself.

However, for some reason, when I upload a file, the calculated checksum is different from the actual checksum I get when I calculate it locally.

I have calculated the checksum locally of the uploaded file (`md5 path/to/file` on OS X) and compared it to the following situations:

- When I upload the file directly to S3 using their Management Console, the checksum matches what I calculate locally
- When I upload the file using skipper-s3, the checksum is different (via `curl -F track=@/path/to/file http://path-to-server`)

The invalid checksum appears for every file I upload so it does not seem to be content-dependent.

As a temporary workaround, is there any possibility to disable sending the `content-md5` header to S3 during upload?

Thanks for your help! Let me know if I can provide any troubleshooting information.

PS. Here is the code I use to upload the file. It looks very ordinary to me. :)

```js
function upload (req, res) {

  var eventid = req.param('eventid')
    , trackid = req.param('trackid')

  req.file('track').upload(
    { adapter: require('skipper-s3')
    , key: sails.config.aws.accessKeyId
    , secret: sails.config.aws.secretAccessKey
    , bucket: sails.config.s3.params.Bucket
    , dirname: eventid
    , saveAs: trackid + '.ogg'
    , maxBytes: 100 * 1024 * 1024   // 100 MB
    }
  , function (err, files) {
      if (err)
        return res.serverError(err)

      return res.ok(files)
    }
  )
}
```

**Comment by sgress454** [comment]
Is the server rejecting your files?

**Comment by Alaneor** [comment]
No, the server accepts the files and also takes the provided MD5 checksum with it, even though it is incorrect...

It seems that S3 simply trusts in the validity of the checksum and does not validate it itself. There is no indication whatsoever that the checksum might be invalid, the issue only appears when I download the file, calculate the checksum of that file and compare it with the ETag header S3 provides.

Which indicates the file has been uploaded correctly, just the checksum has been miscalculated.

Note that when no content-md5 header is provided, S3 calculates the checksum itself.

**Comment by Alaneor** [comment]
Is there anything I can do to help further this issue...? I was unable to track down the issue in the code so I am afraid I cannot help with a patch, but if there is troubleshooting to be done let me know!

Thanks for your help!

**Comment by sailsbot** [comment]
Thanks for posting, @Alaneor.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
